sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -610,8 +796,37 @@ protected void finishWrite() throws Exception {}
      * Opens the channel.
      */
     @Override
-    public final void open(String uId) throws Exception {
+    public final void openWindowed(String uId,
+                                   BoundedWindow window,
+                                   PaneInfo paneInfo,
+                                   int shard,
+                                   int numShards) throws Exception {
+      if (!getWriteOperation().windowedWrites) {
+        throw new IllegalStateException("openWindowed called a non-windowed sink.");
+      }
+      open(uId, window, paneInfo, shard, numShards);
+    }
+
+    @Override
+    public final void openUnwindowed(String uId,
+                                     int shard,
+                                     int numShards) throws Exception {
+      if (getWriteOperation().windowedWrites) {
+        throw new IllegalStateException("openUnwindowed called a windowed sink.");
+      }
+      open(uId, null, null, shard, numShards);
+    }
+
+    private void open(String uId,
```

Comment:
Don't think so, because in the case of dynamic sharding (which will be more full supported in the followup PR), the filename cannot be calculated until finalize.